File: GithubBrowserSample2/app/src/java/java/com/android/example/github/AppExecutors.java
package com.android.example.github;

import android.os.Handler;
import android.os.Looper;

import org.jetbrains.annotations.NotNull;

import java.util.concurrent.Executor;

import javax.inject.Singleton;

import androidx.annotation.NonNull;

/**
 * @author chenweiming
 * @version 1.0
 * @mail cwm930215@gmail.com
 * @since 2018/10/29 9:29 AM
 **/
@Singleton
public class AppExecutors {

    private final Executor diskIO;
    private final Executor networkIO;
    private final Executor mainThread;

    @NotNull
    public final Executor diskIO() {
        return this.diskIO;
    }

    @NotNull
    public final Executor networkIO() {
        return this.networkIO;
    }

    @NotNull
    public final Executor mainThread() {
        return this.mainThread;
    }

    public AppExecutors(@NotNull Executor diskIO,@NotNull Executor networkIO,@NotNull Executor mainThread){
        super();
        this.diskIO = diskIO;
        this.networkIO = networkIO;
        this.mainThread = mainThread;
    }




    public static class MainThreadExecutor implements Executor {
        private Handler mainThreadHandler = new Handler(Looper.getMainLooper());

        @Override
        public void execute(@NonNull Runnable command) {
            mainThreadHandler.post(command);
        }
    }


}

File: GithubBrowserSample3/app/src/main/java/com/android/example/github/AppExecutors.kt
package com.android.example.github

import android.os.Handler
import android.os.Looper
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import javax.inject.Inject
import javax.inject.Singleton

/**
 * @author chenweiming
 * @version 1.0
 * @mail cwm930215@gmail.com
 * @since 2019/3/5 6:28 PM
 **/
@Singleton
open class AppExecutors(
        private val diskIO: Executor,
        private val networkIO: Executor,
        private val mainThread: Executor) {

    @Inject
    constructor() : this(
            Executors.newSingleThreadExecutor(),
            Executors.newFixedThreadPool(3),
            MainThreadExecutor()
    )

    fun diskIO(): Executor {
        return diskIO
    }

    fun networkIO(): Executor {
        return networkIO
    }

    fun mainThread(): Executor {
        return mainThread
    }

    private class MainThreadExecutor : Executor {
        override fun execute(command: Runnable) {
            mainThreadHandler.post(command)
        }

        private val mainThreadHandler = Handler(Looper.getMainLooper())


    }


}